### PR TITLE
Allow to send custom message in `checkError`

### LINKED ISSUE #6547: Allow to send custom message in `checkError`
**Is your feature request related to a problem? Please describe.**
In `checkAuth` it is possible to send a custom message `message: string`
https://github.com/marmelab/react-admin/blob/d33a4ee57e1e6347c9f9ad435b197c00af2d6e3a/docs/Authentication.md#L609
https://github.com/marmelab/react-admin/blob/d33a4ee57e1e6347c9f9ad435b197c00af2d6e3a/packages/ra-core/src/auth/useCheckAuth.ts#L67-L70

I would like prefer the same in `checkError`. Than is it possible to show in dependency of the error the matching message.
At the moment the message is fixed and you can only disable it.
https://github.com/marmelab/react-admin/blob/d33a4ee57e1e6347c9f9ad435b197c00af2d6e3a/docs/Authentication.md#L608
https://github.com/marmelab/react-admin/blob/d33a4ee57e1e6347c9f9ad435b197c00af2d6e3a/packages/ra-core/src/auth/useLogoutIfAccessDenied.ts#L73-L82

**Describe the solution you'd like**
I would like to send custom messages in `checkError` like:
```js
  checkError: ({ status }) => {
    if (status === 401) {
      return Promise.reject({ message: "Message 401" });
    } else if (status === 403) {
      return Promise.reject({ message: "Message 403" });
    }
    return Promise.resolve();
  },
```

### PR DESCRIPTION
**Pull request by dklimpel**
Fixes #6547

Allow to send custom message in `checkError` like in `checkAuth`
https://github.com/marmelab/react-admin/blob/d33a4ee57e1e6347c9f9ad435b197c00af2d6e3a/docs/Authentication.md#L609
https://github.com/marmelab/react-admin/blob/d33a4ee57e1e6347c9f9ad435b197c00af2d6e3a/packages/ra-core/src/auth/useCheckAuth.ts#L67-L70

Example
```js
  checkError: ({ status }) => {
    if (status === 401) {
      return Promise.reject({ message: "Message 401" });
    } else if (status === 403) {
      return Promise.reject({ message: "Message 403" });
    }
    return Promise.resolve();
  },
```

### CODE REVIEW

**Comment by fzaninotto** [comment]
Thanks for the PR! Could you add a unit test?

**Comment by dklimpel** [comment]
Im not sure how to creat a good unit test for it.
At the moment there is a test:
https://github.com/marmelab/react-admin/blob/b0f997473868c68746188ef92ba2e325974f3ec5/packages/ra-core/src/auth/useLogoutIfAccessDenied.spec.tsx#L108-L119

But it checks only that there is one notification, not the content of it.

I could not find a similar unit test in https://github.com/marmelab/react-admin/blob/next/packages/ra-core/src/auth/useCheckAuth.spec.tsx

**Comment by dklimpel** [comment]
Is this a good example for check the notify message?
https://github.com/marmelab/react-admin/blob/43c4fafc8bcedebc386c7d3dc3b63cfd56420a17/packages/ra-core/src/controller/details/useCreateController.spec.tsx#L122-L132

**Comment by djhi** [comment]
Hi @dklimpel, sorry for the delay. Yes, you can use the same pattern as in `useCreateController.spec.tsx`.

**Comment by dklimpel** [comment]
I am not able to write a test, sorry.
The "problem" is, that there is no `dispatch`.
Similar places in the code test the notifications in this `dispatch`
E.g.
https://github.com/marmelab/react-admin/blob/43c4fafc8bcedebc386c7d3dc3b63cfd56420a17/packages/ra-core/src/auth/Authenticated.spec.tsx#L51-L56

https://github.com/marmelab/react-admin/blob/43c4fafc8bcedebc386c7d3dc3b63cfd56420a17/packages/ra-core/src/controller/details/useCreateController.spec.tsx#L119-L121

This test is valid, bot has no test for notification, because `expect(dispatch).toHaveBeenCalledTimes(0);` 
```js
    it('new test', async () => {
        const history = createMemoryHistory();

        const { dispatch, queryByText } = renderWithRedux(
            <Router history={history}>
	            <AuthContext.Provider value={authProvider}>
	                <TestComponent error={new Error('denied')} />
	                <TestComponent error={new Error('denied')} />
	            </AuthContext.Provider>
            </Router>
        );
        await waitFor(() => {
            expect(authProvider.logout).toHaveBeenCalledTimes(1);
            expect(notify).toHaveBeenCalledTimes(1);
            expect(dispatch).toHaveBeenCalledTimes(0);
            expect(queryByText('logged in')).toBeNull();
        });
    });
```

**Comment by WiXSL** [comment]
> I am not able to write a test, sorry. The "problem" is, that there is no `dispatch`. Similar places in the code test the notifications in this `dispatch` E.g.
> 

@dklimpel, you have access to the notifications, you just have to modify the `authProvider.checkAuth` function  inside `useLogoutIfAccessDenied.spec.tsx` to show any message.
You can do it like this:

```js
checkError: params => {
      if (params instanceof Error) {
          return Promise.reject(new Error(params.message === 'denied' ? 'logout' : params.message));
      }
      return Promise.resolve();
  },
```
And then make a test like this:

```jsx
    it('should notify if passed an error with a message that makes the authProvider throw', async () => {
        const { queryByText } = renderInRouter(
            <AuthContext.Provider value={authProvider}>
                <TestComponent error={new Error('Test message')} />
            </AuthContext.Provider>
        );
        await waitFor(() => {
            expect(authProvider.logout).toHaveBeenCalledTimes(1);
            expect(notify).toHaveBeenCalledTimes(1);
            expect(notify.mock.calls[0][0]).toEqual('Test message');
            expect(queryByText('logged in')).toBeNull();
        });
    });

**Comment by WiXSL** [comment]
Can we dismiss the error CodeQL is throwing?

**Comment by dklimpel** [comment]
Can you re-run CI? I have merged newest `next` branch.

**Comment by dklimpel** [comment]
I am confused why there is an error in `navigation.js`

```
  Running:  navigation.js                                                                  (7 of 10)


  Navigation
    Sidebar
      1) should have tabbable menu items
    Skip Navigation Button
      ✓ should appear when a user immediately tabs on the homepage (307ms)


  1 passing (5s)
  1 failing

  1) Navigation
       Sidebar
         should have tabbable menu items:
     AssertionError: Timed out retrying: expected '<a.MuiMenuItem-root.MuiMenuItem-gutters.MuiButtonBase-root.RaMenuItemLink-root.css-is19is.RaMenuItemLink-active>' to have class 'Mui-focusVisible'
      at Context.eval (http://localhost:8080/__cypress/tests?p=integration/navigation.js:111:66)
```

**Comment by WiXSL** [comment]
I've re-run it, now seems ok

**Review by WiXSL — APPROVED** [review]

File: docs/Authentication.md [hunk 1]
```diff
@@ -614,7 +614,7 @@ When the auth backend returns an error, the Auth Provider should return a reject
 | Method           | Reject if                                 | Error format |
 | ---------------- | ----------------------------------------- | --------------- |
 | `login`          | Login credentials weren't accepted        | `string | { message?: string }` error message to display |
-| `checkError`     | Error is an auth error                    | `void | { redirectTo?: string, message?: boolean }` route to redirect to after logout, and whether to disable error notification |
+| `checkError`     | Error is an auth error                    | `void | { redirectTo?: string, message?: string }` route to redirect to after logout, message to notify the user |
```
Comment (WiXSL) [inline docs/Authentication.md:617]:
```suggestion
| `checkError`     | Error is an auth error                    | `void | { redirectTo?: string, message?: string | boolean  }` route to redirect to after logout, message to notify the user or `false` to disable notification |
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Review by fzaninotto — APPROVED** [review]

**Comment by fzaninotto** [comment]
Thanks!

**Pull request merged**